datafusion/optimizer/src/optimize_projections.rs

```diff
@@ -162,14 +162,36 @@ fn optimize_projections(
                 .map(|input| ((0..input.schema().fields().len()).collect_vec(), false))
                 .collect::<Vec<_>>()
         }
+        LogicalPlan::Extension(extension) => {
+            let necessary_children_indices = if let Some(necessary_children_indices) =
+                extension.node.necessary_children_exprs(indices)
+            {
+                necessary_children_indices
+            } else {
+                // Requirements from parent cannot be routed down to user defined logical plan safely
+                return Ok(None);
+            };
+            let children = extension.node.inputs();
+            assert_eq!(children.len(), necessary_children_indices.len());
```

Comment:
Can we make this an internal error instead of a panic, to try and provide a nicer failure mode?